src/Chapter3.hs

```diff
@@ -1006,6 +1120,55 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
+data KnightAction = KAttack | Spell | Drink deriving (Eq,Enum)
+data MonsterAction = MAttack | RunAway deriving (Eq,Enum)
+
+data MyKnight = MyKnight
+  { knightHealth :: Health
+  , knightAttack :: Attack
+  , knightDefense :: Defense
+  , knightActions :: [KnightAction]
+  }
+data MyMonster = MyMonster
+  { monsterHealth :: Health
+  , monsterAttack :: Attack
+  , monsterActions :: [MonsterAction]
+  }
+
+class Action a where
+  act :: [a] -> (a,[a])
+
+instance Action KnightAction where
+  act (x:xs) = (x,xs)
+instance Action MonsterAction where
+  act (x:xs) = (x,xs)
+
+class TypeFighter a where
+  _attack :: a -> Attack
+
+instance TypeFighter MyKnight where
+  _attack :: MyKnight -> Attack
+  _attack k = knightAttack k
+
+instance TypeFighter MyMonster where
+  _attack :: MyMonster -> Attack
+  _attack m = monsterAttack m
+
+hit :: Attack -> Health -> Maybe Defense -> Health
+hit (Attack a) (Health h) (Just (Defense d)) = Health (h + d - a)
+hit (Attack a) (Health h) Nothing = Health (h - a)
+
+drink :: MyKnight -> MyKnight
+drink k = k{knightHealth = Health (h + 50) } where
+  (Health h) = knightHealth k
+
+spell :: MyKnight -> MyKnight
+spell k = k{knightDefense =  Defense (d + 25) } where
+  (Defense d) = knightDefense k
+
+runaway :: MyMonster -> MyMonster
+runaway m = m{monsterHealth = Health 0}
+
```

Comment:
This is the great stuff that you've created in here!
All this is parts of the fight function, and can be easily composed together in one function :+1: